nextest-runner/src/test_output.rs

```diff
@@ -0,0 +1,322 @@
+    #[inline]
+    pub fn stderr(&mut self) -> TestOutputWriter<'_> {
+        TestOutputWriter {
+            acc: self,
+            stdout: false,
+        }
+    }
+}
+
+/// Provides [`std::io::Write`] and [`std::fmt::Write`] implementations for a
+/// single stream of a [`TestOutputAccumulator`]
+pub struct TestOutputWriter<'acc> {
+    acc: &'acc mut TestOutputAccumulator,
+    stdout: bool,
+}
+
+impl<'acc> std::io::Write for TestOutputWriter<'acc> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let start = self.acc.buf.len();
+        self.acc.buf.extend_from_slice(buf);
+        self.acc.chunks.push(OutputChunk {
+            range: start..self.acc.buf.len(),
+            timestamp: Instant::now(),
+            stdout: self.stdout,
+        });
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl<'acc> std::fmt::Write for TestOutputWriter<'acc> {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        match self.write(s.as_bytes()) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(std::fmt::Error),
+        }
+    }
+}
+
+pin_project! {
+    #[project = StreamProj]
+    enum Stream {
+        Stdout { #[pin] o: tokio::io::BufReader<tokio::process::ChildStdout> },
+        Stderr { #[pin] o: tokio::io::BufReader<tokio::process::ChildStderr> },
+    }
+}
+
+impl tokio::io::AsyncRead for Stream {
+    #[inline]
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        match self.project() {
+            StreamProj::Stdout { o } => o.poll_read(cx, buf),
+            StreamProj::Stderr { o } => o.poll_read(cx, buf),
+        }
+    }
+}
+
+impl tokio::io::AsyncBufRead for Stream {
+    #[inline]
+    fn poll_fill_buf(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::io::Result<&[u8]>> {
+        match self.project() {
+            StreamProj::Stdout { o } => o.poll_fill_buf(cx),
+            StreamProj::Stderr { o } => o.poll_fill_buf(cx),
+        }
+    }
+
+    #[inline]
+    fn consume(self: Pin<&mut Self>, amt: usize) {
+        match self.project() {
+            StreamProj::Stdout { o } => o.consume(amt),
+            StreamProj::Stderr { o } => o.consume(amt),
+        }
+    }
+}
+
+/// The size of each buffered reader's buffer, and the size at which we grow
+/// the interleaved buffer.
+///
+/// This size is not totally arbitrary, but rather the (normal) page size on
+/// most linux, windows, and macos systems.
+const CHUNK_SIZE: usize = 4 * 1024;
+
+/// Collects the stdout and/or stderr streams into a single buffer
+#[allow(clippy::needless_lifetimes, clippy::let_and_return)]
+pub fn collect_test_output<'a>(
+    stdout: Option<tokio::process::ChildStdout>,
+    stderr: Option<tokio::process::ChildStderr>,
```

Comment:
Can the API be changed to `Option<(ChildStdout, ChildStderr)>` to avoid the `unreachable!()` used in match?